source/validate_cfg.cpp

```diff
@@ -244,30 +254,154 @@ spv_result_t FirstBlockAssert(ValidationState_t& _, uint32_t target) {
 }
 
 spv_result_t MergeBlockAssert(ValidationState_t& _, uint32_t merge_block) {
-  if (_.get_current_function().IsMergeBlock(merge_block)) {
+  if (_.get_current_function().IsBlockType(merge_block, kBlockTypeMerge)) {
     return _.diag(SPV_ERROR_INVALID_CFG)
            << "Block " << _.getIdName(merge_block)
            << " is already a merge block for another header";
   }
   return SPV_SUCCESS;
 }
 
+/// Update the continue construct's exit blocks once the backedge blocks are
+/// identified in the CFG.
+void UpdateContinueConstructExitBlocks(
+    Function& function, const vector<pair<uint32_t, uint32_t>>& back_edges) {
+  auto& constructs = function.get_constructs();
+  // TODO(umar): Think of a faster way to do this
+  for (auto& edge : back_edges) {
+    uint32_t back_edge_block_id;
+    uint32_t loop_header_block_id;
+    tie(back_edge_block_id, loop_header_block_id) = edge;
+
+    auto is_this_header = [=](Construct& c) {
+      return c.get_type() == ConstructType::kLoop &&
+             c.get_entry()->get_id() == loop_header_block_id;
+    };
+
+    for (auto construct : constructs) {
+      if (is_this_header(construct)) {
+        Construct* continue_construct =
+            construct.get_corresponding_constructs().back();
+        assert(continue_construct->get_type() == ConstructType::kContinue);
+
+        BasicBlock* back_edge_block;
+        tie(back_edge_block, ignore) = function.GetBlock(back_edge_block_id);
+        continue_construct->set_exit(back_edge_block);
+      }
+    }
+  }
+}
+
+/// Constructs an error message for construct validation errors
+string ConstructErrorString(Construct construct, string header_string,
+                            string exit_string, bool post_dominate = false) {
+  string construct_name;
+  string header_name;
+  string exit_name;
+  string dominate_text = "dominate";
+  if (post_dominate) {
+    dominate_text = "post " + dominate_text;
+  }
+  switch (construct.get_type()) {
+    case ConstructType::kSelection:
+      construct_name = "selection";
+      header_name = "selection header";
+      exit_name = "merge block";
+      break;
+    case ConstructType::kLoop:
+      construct_name = "loop";
+      header_name = "loop header";
+      exit_name = "merge block";
+      break;
+    case ConstructType::kContinue:
+      construct_name = "continue";
+      header_name = "continue target";
+      exit_name = "back-edge block";
+      break;
+    case ConstructType::kCase:
+      construct_name = "case";
+      header_name = "case block";
+      exit_name = "exit block";  // TODO(umar): there has to be a better name
+      break;
+    default:
+      assert(1 == 0 && "Not defined type");
+  }
+  // TODO(umar): Add header block for continue constructs to error message
+  return "The " + construct_name + " construct with the " + header_name + " " +
```

Comment:
In the post-dominate case, this reads incorrectly, e.g. I get:

The continue construct with the continue target 7 does not post dominate the back-edge block 10

That middle text should be "is not post dominated by"